readme.txt
ufw2web is a web interface for ubuntu-firewall (ufw). Right now, the only 
option for ufw2web, is to enable or disable the firewall. 

Configuration of ufw it self, must be done via the approriate commands
on the command line. See https://wiki.ubuntu.com/UbuntuFirewall and 'man ufw'
for more details.

SYSTEM REQUIREMENTS
- Ubuntu Linux: Hardy 8.04 (server or desktop) or higher
  - on these systems, ufw is installed by default

INSTALLATION
- install CherryPy3 and ConfigObj:
  sudo apt-get install python-cherrypy3 python-configobj
- copy files of ufw2web to a local location, preferable: /usr/local/bin/ufw2web
- test: start webserver with sudo (root) privileges
  (sudo) python ufw2web.py
  - exit: Ctrl-C
- autostart on system boot via /etc/rc.local:
  - add the following line in /etc/rc.local (just before 'exit 0'):
    python <location of ufw2web files>/ufw2web.py

CONFIGURATION
- defaults:
  - WebServerPort: 80
  - UseAuthentication: 0 (off)
  - LoginUser/LoginPassword: admin/admin
    - not used when UseAuthentication is off
- change defaults:
  - uncomment keys in config/ufw2web.conf and place the desired values

TRANSLATIONS
- by default, language is English
- all strings can be translated by ufw2web-translations.conf
  - placed in config/
- provided are two files: 
  - ufw2web-translations-nl.conf (out-of-the-box Dutch translation)
  - ufw2web-translations-template.conf (base file for any other language)
- just rename one of them to ufw2web-translations.conf and restart the server



changelog.txt
0.3
  - Added favicon
  - Changed readme.txt: ufw2web is dependent from python-cherrypy3 and
    not python-cherrypy (actually: version 2.x)
  - Added extra config parameter: 'server.socket_host':'0.0.0.0'
    In some cases, the server was not reachable from the outside
  - Simplified status checking: see if 'not loaded' is in the return string
    This was needed to obtain compatability with Hardy, which is using on older
    version of ufw. Status returned by Hardy was 'Firewall not loaded' instead 
    of 'Status: not loaded' in Intrepid and higher.
  - ConfigFolder is now based defined as an absolute path to provide startup
    in an init-system

0.2 
  - Added authentication, default admin/admin
    - highly configurable, see config/ufw2web.conf for details
  - Changed structure of ufw2web.conf, not containing directory configuration anymore
    - refacored (reading of) settings
  - Autoredirect to main page for result pages (removed OK button)
  - Moved all config files to ./config (including translations)
  - Reading config files is more robust when keys do not exist
  - Use string formatting (code refactoring)

0.1 First public version


ufw2web.py
#!/usr/bin/python
#
#  uwf2web: Web interface for UFW (Ubuntu FireWall)
#
#  Usage: (sudo) python uwf2web
#
#    Copyright (C) 2009 Mark Baaijens <mark.baaijens@gmail.com>
#
#    This file is part of uwf2web.
#
#    uwf2web is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    uwf2web is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
import cherrypy  # From package python-cherrypy
import subprocess
from string import *
import os.path
import os
from configobj import ConfigObj # From package python-configobj

# UFW status codes
usUnknown = -1 
usDisabled = 0  
usEnabled = 1  

# Static page segments
_header = '''<html>
  <head>
    <title>Web interface for Ubuntu Firewall (UFW)</title>
    <link rel="icon" type="image/png" href="images/icon.png">
  </head>
  <body>
  <div class="container">'''

_footer = '''
  </div>
  </body>
</html>'''

# Some constants
ConfigFolder = os.path.dirname( os.path.realpath( __file__ )) + '/config'
Version = '0.3'

class Start:
  def index(self):
    text = _header
    ufwStatus = self.uwfStatus()
    if ufwStatus == usDisabled:
      text = text + '''
          <img src="images/disabled.png" align="left">
          <h2>%s</h2>
          <p>
          <form action="enablefirewall" method="GET">
          %s<p>
          <input type="submit" value="%s"/>
          </form>''' % (strFirewallStatusDisabled, strSystemUnprotected, strEnableFirewall)

    elif ufwStatus == usEnabled:
      text = text + '''
          <img src="images/enabled.png" align="left">
          <h2>%s</h2>
          <p>
          <form action="disablefirewall" method="GET">
          %s<p>
          <input type="submit" value="%s"/>
          </form>''' % (strFirewallStatusEnabled, strSystemProtected, strDisableFirewall)

    elif ufwStatus == usUnknown:
      text = text + '''
          <img src="images/unknown.png" align="left">
          <h2>%s</h2>
          <p>
          %s
          <p>''' % (strFirewallStatusUnknown, strIsSudoPrivileges)

    text = text + _footer
    return text
  index.exposed = True

  def enablefirewall(self):
    process = subprocess.Popen(["ufw","enable"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)

    # Page will be redirected after a timeout to the main page
    text = '''
        <html>
        <head>
        <meta http-equiv="refresh" content="2; url=./">
        </head>
        <body>%s</body>
        <html>''' % strFirewallIsEnabled
    return text
  enablefirewall.exposed = True  

  def disablefirewall(self):
    process = subprocess.Popen(["ufw","disable"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)

    # Page will be redirected after a timeout to the main page
    text = '''
        <html>
        <head>
        <meta http-equiv="refresh" content="2; url=./">
        </head>
        <body>%s</body>
        <html>''' % strFirewallIsDisabled
    return text
  disablefirewall.exposed = True  

  def uwfStatus(self):
    process = subprocess.Popen(["ufw","status"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    output = process.communicate()[0]

    # Apparently, 'ufw status' is trailed by an invisible character, so it is impossible
    # to do a direct string comparison with '=='; that's why only the 'in' operator is 
    # working here
    if 'not loaded' in output:
      result = 0
    elif 'loaded' in output: # Just to make sure there *is* some output
      result = 1
    else:
      # Probably no sudo privileges: these are essential, even for 'ufw status'
      result = -1  
    return result

# Output to console
print 'ufw2web version %s' % Version
print '- Copyright (C) 2009 Mark Baaijens'
print '- License GNU GPL version 3 or later'

#
# Translations
#

# Translation strings (default)
strFirewallStatusDisabled = 'Firewall (UFW) is disabled.'
strFirewallStatusEnabled = 'Firewall (UFW) is enabled.'
strFirewallStatusUnknown = 'Firewall (UFW) status is unknown.'
strFirewallIsDisabled = 'Firewall (UFW) is deactivated.'
strFirewallIsEnabled = 'Firewall (UFW) is activated.'
strSystemUnprotected = 'The system is unprotected.'
strSystemProtected = 'The system is protected.'
strEnableFirewall = 'Enable firewall'
strDisableFirewall = 'Disable firewall'
strIsSudoPrivileges = 'Is ufw2web.py running with sudo privileges?'

# Translations can be done externally, by the file ufw2web-translations.conf.
TranslationsFile = ConfigFolder + "/ufw2web-translations.conf"
if os.path.isfile(TranslationsFile):
  print 'Reading translations from ' + TranslationsFile
  config = ConfigObj(TranslationsFile)
  try:
    strFirewallStatusDisabled = config['FirewallStatusDisabled']
    strFirewallStatusEnabled = config['FirewallStatusEnabled']
    strFirewallStatusUnknown = config['FirewallStatusUnknown']
    strFirewallIsDisabled = config['FirewallIsDisabled']
    strFirewallIsEnabled = config['FirewallIsEnabled']
    strSystemUnprotected = config['SystemUnprotected']
    strSystemProtected = config['SystemProtected']
    strEnableFirewall = config['EnableFirewall']
    strDisableFirewall = config['DisableFirewall']
    strIsSudoPrivileges = config['IsSudoPrivileges']
  except KeyError:
    print '- Warning: one or more keys not found in %s' % TranslationsFile
else:
  print 'No translation file %s, using defaults (English)' % TranslationsFile

#
# Server configuration
#

# Default settings
WebServerPort = 80
UseAuthentication = 0
LoginUser = 'admin'
LoginPassword = 'admin'

# Read settings from config file (if existing)
ConfigFile = ConfigFolder + '/ufw2web.conf'
if os.path.isfile(ConfigFile):
  print 'Reading settings from ' + ConfigFile
  config = ConfigObj(ConfigFile)

  # Retrieve settings
  try:
    WebServerPort = config['WebServerPort']
    try:
      WebServerPort = int(WebServerPort)
    except ValueError:
      print '- Warning: no valid value for WebServerPort'
    print '- WebServerPort: %d' % WebServerPort
  except KeyError:
    print '- WebServerPort not found, using default: %d' % WebServerPort

  try:
    UseAuthentication = config['UseAuthentication']
    try:
      UseAuthentication = int(UseAuthentication)
    except ValueError:
      print '- Warning: no valid value for UseAuthentication'
    print '- UseAuthentication: %d' % UseAuthentication
  except KeyError:
    print '- UseAuthentication not found, using default: %d' % UseAuthentication

  # If there's no authentication, there is no use for user credentials
  if (UseAuthentication == 1):
    try:
      LoginUser = config['LoginUser']
      print '- LoginUser: %s' % LoginUser
    except KeyError:
      print '- LoginUser not found, using default: %s' % LoginUser

    try:
      LoginPassword = config['LoginPassword']
      print '- LoginPassword: %s' % LoginPassword
    except KeyError:
      print '- LoginPassword not found, using default: %s' % LoginPassword

# Basic server configuration
current_dir = os.path.dirname(os.path.abspath(__file__))
cherrypy.config.update({'server.socket_port':WebServerPort,
                        'server.socket_host':'0.0.0.0',
                        'tools.staticdir.root':current_dir}) # Root for static folders

# User dictionairy for authentication
users = {LoginUser: LoginPassword}
        
# Extra config       
conf = {'/images': {'tools.staticdir.on': True,    # Folder 'images' is a static folder
                    'tools.staticdir.dir': 'images'},
        '/': {'tools.digest_auth.on': (UseAuthentication == 1), # Activate authentication (on demand)
              'tools.digest_auth.realm': '',
              'tools.digest_auth.users': users}}
#
# And start the webserver...
#
cherrypy.quickstart(Start(), '/', config=conf)

